plutus-conformance/src/PlutusConformance/Common.hs

```diff
@@ -68,6 +74,44 @@ checkExpected (AcceptTests accept) expectedFile actual = do
             pure $ testPassed "Unexpected output, accepted it"
         else pure $ testFailed $ "Unexpected output:" ++ show actual
 
+{- | The default shown text when a parse error occurs.
+We don't want to show the detailed parse errors so that
+users of the test suite can produce this expected outputs more easily. -}
+shownParseError :: T.Text
+shownParseError = "parse error"
+
+-- | The default shown text when evaluation fails.
+shownEvaluationFailure :: T.Text
+shownEvaluationFailure = "evaluation failure"
+
+-- UPLC evaluation test functions
+
+type UplcProg = UPLC.Program Name DefaultUni DefaultFun ()
+
+type UplcEvaluator = UplcProg -> Maybe UplcProg
+
+-- | A UPLC evaluation test suite.
+data UplcEvaluationTest =
+    MkUplcEvaluationTest {
+       -- | The evaluator function to use for the test.
+       evaluator :: UplcEvaluator
+       -- | The test directory in which the test files are located.
+       , testDir :: FilePath
+    }
+
+-- | Our `evaluator` for the Haskell UPLC tests is the CEK machine.
+evalUplcProg :: UplcEvaluator
+evalUplcProg = traverseOf UPLC.progTerm eval
+  where
+    eval t = do
+        -- The evaluator throws if the term has free variables
+        case UPLC.deBruijnTerm t of
+            Left (_ :: UPLC.FreeVariableError) -> Nothing
+            Right _                            -> Just ()
```

Comment:
I'm quite confused: the CEK machine allows one to throw an `OpenTermEvaluatedMachineError`, which we later catch and return in a pure fashion. But in addition to that we also purely return `FreeVariableError`, which we later throw? I.e. we have two ways of returning the same error (granted, from two different parts of the machinery, but still), which already seems excessive, but then we handle them in opposite ways and it's just weird.

@bezirg @michaelpj am I misunderstanding it? Should we do it differently?

Reply:
It's the handling of open terms. The evaluation functions check that the term is not open, and throws if it is. I actually don't remember why we didn't just return the corresponding machine exception :thinking:

Reply:
prod @bezirg